File: .migration-backup/artifacts/api-server/src/lib/parser.ts
// Best-effort text extraction for common knowledge-base formats.
// Heavy parsers (pdf-parse, mammoth, xlsx) are dynamically imported so a
// missing or broken module never crashes the whole server.

export type ParsedFile = { text: string; type: string };

export async function extractText(
  buffer: Buffer,
  filename: string,
  mimetype: string,
): Promise<ParsedFile> {
  const lower = filename.toLowerCase();
  const ext = lower.includes(".") ? lower.split(".").pop()! : "";

  if (ext === "txt" || ext === "md" || mimetype.startsWith("text/")) {
    return { text: buffer.toString("utf8"), type: ext || "txt" };
  }

  if (ext === "pdf" || mimetype === "application/pdf") {
    try {
      // pdf-parse default export: (buf) => { text }
      const mod = (await import("pdf-parse")) as unknown as {
        default?: (b: Buffer) => Promise<{ text: string }>;
        pdf?: (b: Buffer) => Promise<{ text: string }>;
      };
      const fn = mod.default || mod.pdf;
      if (!fn) return { text: "", type: "pdf" };
      const out = await fn(buffer);
      return { text: out.text, type: "pdf" };
    } catch {
      return { text: "", type: "pdf" };
    }
  }

  if (ext === "docx" || mimetype.includes("officedocument.wordprocessingml")) {
    try {
      const mod = (await import("mammoth")) as {
        extractRawText: (input: { buffer: Buffer }) => Promise<{ value: string }>;
      };
      const out = await mod.extractRawText({ buffer });
      return { text: out.value, type: "docx" };
    } catch {
      return { text: "", type: "docx" };
    }
  }

  if (ext === "xlsx" || ext === "xls" || mimetype.includes("spreadsheetml")) {
    try {
      const xlsx = (await import("xlsx")) as typeof import("xlsx");
      const wb = xlsx.read(buffer, { type: "buffer" });
      const parts: string[] = [];
      for (const sheetName of wb.SheetNames) {
        const sheet = wb.Sheets[sheetName];
        if (!sheet) continue;
        parts.push(`# ${sheetName}\n${xlsx.utils.sheet_to_csv(sheet)}`);
      }
      return { text: parts.join("\n\n"), type: "xlsx" };
    } catch {
      return { text: "", type: "xlsx" };
    }
  }

  // Fallback: treat as utf-8 text.
  return { text: buffer.toString("utf8"), type: ext || "bin" };
}

// Chunk text into ~800-char segments on whitespace boundaries.
export function chunkText(text: string, target = 800): string[] {
  const cleaned = text.replace(/\r/g, "").replace(/\u0000/g, "").trim();
  if (!cleaned) return [];

  const chunks: string[] = [];
  let buf = "";
  const paragraphs = cleaned.split(/\n{2,}/);

  for (const para of paragraphs) {
    const piece = para.trim();
    if (!piece) continue;
    if ((buf + "\n\n" + piece).length <= target) {
      buf = buf ? `${buf}\n\n${piece}` : piece;
      continue;
    }
    if (buf) chunks.push(buf);
    if (piece.length <= target) {
      buf = piece;
    } else {
      // Hard-split very long paragraphs.
      buf = "";
      for (let i = 0; i < piece.length; i += target) {
        chunks.push(piece.slice(i, i + target));
      }
    }
  }
  if (buf) chunks.push(buf);
  return chunks;
}

File: artifacts/api-server/app/services/parser.py
import io
import re


def _chunk_text(text: str, target: int = 800) -> list[str]:
    paragraphs = re.split(r"\n\n+", text)
    chunks: list[str] = []
    current = ""
    for para in paragraphs:
        para = para.strip()
        if not para:
            continue
        if len(para) > target:
            # Hard-split long paragraphs on word boundaries
            words = para.split()
            buf = ""
            for word in words:
                if len(buf) + len(word) + 1 > target and buf:
                    chunks.append(buf.strip())
                    buf = word
                else:
                    buf = (buf + " " + word).strip()
            if buf:
                chunks.append(buf)
            continue
        if len(current) + len(para) + 2 > target and current:
            chunks.append(current.strip())
            current = para
        else:
            current = (current + "\n\n" + para).strip() if current else para
    if current:
        chunks.append(current.strip())
    return [c for c in chunks if c]


async def extract_text(content: bytes, filename: str, content_type: str) -> tuple[str, str]:
    name_lower = filename.lower()

    if content_type.startswith("text/") or name_lower.endswith((".txt", ".md")):
        return content.decode("utf-8", errors="replace"), "text"

    if content_type == "application/pdf" or name_lower.endswith(".pdf"):
        from pypdf import PdfReader
        reader = PdfReader(io.BytesIO(content))
        text = "\n".join(page.extract_text() or "" for page in reader.pages)
        return text, "pdf"

    if name_lower.endswith(".docx") or content_type in (
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    ):
        import docx
        doc = docx.Document(io.BytesIO(content))
        text = "\n".join(p.text for p in doc.paragraphs)
        return text, "docx"

    if name_lower.endswith((".xlsx", ".xls")) or content_type in (
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        "application/vnd.ms-excel",
    ):
        import openpyxl
        wb = openpyxl.load_workbook(io.BytesIO(content), data_only=True)
        parts = []
        for sheet in wb.worksheets:
            rows = []
            for row in sheet.iter_rows(values_only=True):
                rows.append(",".join(str(c) if c is not None else "" for c in row))
            parts.append("\n".join(rows))
        return "\n\n".join(parts), "xlsx"

    # Fallback: treat as UTF-8 text
    return content.decode("utf-8", errors="replace"), "text"


def chunk_text(text: str, target: int = 800) -> list[str]:
    return _chunk_text(text, target)
